eyelid-match-ops/src/encoded.rs

```diff
@@ -1 +1,218 @@
 //! Iris matching operations on polynomial-encoded bit vectors.
+
+#![allow(
+    clippy::cast_possible_truncation,
+    clippy::cast_possible_wrap,
+    clippy::cast_sign_loss
+)]
+
+use itertools::Itertools;
+use std::error::Error;
+
+use crate::plaintext::{index_1d, IrisCode, IrisMask};
+use crate::primitives::poly::modular_poly::conf::FullRes;
+use crate::primitives::poly::{Coeff, Poly, PolyConf};
+use crate::{
+    IRIS_BIT_LENGTH, IRIS_COLUMNS as NUM_COLS, IRIS_COLUMN_LENGTH as NUM_ROWS,
+    IRIS_MATCH_DENOMINATOR, IRIS_MATCH_NUMERATOR, IRIS_ROTATION_COMPARISONS, IRIS_ROTATION_LIMIT,
+};
+use ark_ff::{One, Zero};
+use num_bigint::BigUint;
+
+#[cfg(any(test, feature = "benchmark"))]
+pub mod test;
+
+// Divide iris codes into blocks that can each fit into a polynomial.
+/// The number of rows in each block.
+const ROWS_PER_BLOCK: usize = 10;
+/// The number of blocks.
+const NUM_BLOCKS: usize = NUM_ROWS / ROWS_PER_BLOCK;
+// Only full blocks are supported at the moment.
+const_assert_eq!(NUM_BLOCKS * ROWS_PER_BLOCK, NUM_ROWS);
+
+// Aliases to match the paper equations.
+/// The number of columns in each block.
+const K: i32 = NUM_COLS as i32;
+/// The right-most rotation.
+const V: i32 = IRIS_ROTATION_LIMIT as i32;
+/// The left-most rotation.
+const U: i32 = -V;
+/// The number of columns plus padding for rotations.
+const DELTA: usize = (K + V - U) as usize;
+/// The number of rows in each block.
+const S: usize = ROWS_PER_BLOCK;
+
+/// An Iris code, encoded in polynomials. To be stored in the database.
+pub struct PolyCode {
+    /// The polynomials, encoding one block of rows each. Storage variant.
+    polys: Vec<Poly<FullRes>>,
```

Comment:
An issue with this `PolyCode` struct definition is that the polynomials used currently are ciphertext polynomials with coefficients mod a prime of size around $2^{79}$, where the plaintexts should properly be polynomials with coefficients mod a 2-power between $2^{10}$ or $2^{15}$.

@emmorais for encryption purposes do we use any algebraic structure associated with plaintext coefficients (e.g. multiply the plaintext polynomials by something as elements in $(\mathbb{Z}/2^k\mathbb{Z})[x]$), or is it really just operations which lift the coefficients directly in some way into the ciphertext polynomial ring?  If it's the latter, is it possible that what we really want is just a vector of coefficients as output for the encoding methods?

Alternatively, if we do need or want a proper polynomial representation of plaintexts, we'll probably need to implement non-field modular arithmetic in terms of operations in the `ark_ff` field interface.

Reply:
> the plaintexts should properly be polynomials with coefficients mod a 2-power between $2^{10}$ or $2^{15}$.

We could define a coefficient type with this range, and use it in a new polynomial config.

But there could be some advantages to testing the ciphertext arithmetic in the same, even if that's not strictly how we'll be using the code in production.

Reply:
In this PR, I isolated this question with the aliases `PlainConf` and `PlainCoeff`.

This could be made configurable after #82; see proposal in [comment](https://github.com/Inversed-Tech/eyelid/pull/82/files#r1599846550).

Reply:
A plaintext is interpreted as a ring polynomial where coefficients are taken modulo 2ˆ10, instead of a prime around 2ˆ79. We are not using arithmetic modulo 2ˆ10 yet, because no homomorphic operation was implemented yet. But next step for us is exactly to provide basic support for this arithmetic. But it should not be too complicated, just reducing coefficients modulo 2ˆ10 will solve it.